File: app/src/main/java/com/certified/notes/adapters/BookMarkRecyclerAdapter.java
package com.certified.notes.adapters;

import android.content.Context;
import android.content.SharedPreferences;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.certified.notes.R;
import com.certified.notes.model.BookMark;
import com.certified.notes.room.NotesViewModel;
import com.certified.notes.util.PreferenceKeys;
import com.like.LikeButton;
import com.like.OnLikeListener;

import java.util.HashSet;
import java.util.Set;

public class BookMarkRecyclerAdapter extends ListAdapter<BookMark, BookMarkRecyclerAdapter.ViewHolder> {

    private static final DiffUtil.ItemCallback<BookMark> DIFF_CALLBACK = new DiffUtil.ItemCallback<BookMark>() {
        @Override
        public boolean areItemsTheSame(@NonNull BookMark oldItem, @NonNull BookMark newItem) {
            return oldItem.getNoteId() == newItem.getNoteId();
        }

        @Override
        public boolean areContentsTheSame(@NonNull BookMark oldItem, @NonNull BookMark newItem) {
            return oldItem.getNoteId() == newItem.getNoteId() &&
                    oldItem.getCourseCode().equals(newItem.getCourseCode()) &&
                    oldItem.getNoteTitle().equals(newItem.getNoteTitle()) &&
                    oldItem.getNoteContent().equals(newItem.getNoteContent());
        }
    };

    private final Context mContext;
    private final NotesViewModel mViewModel;
    private OnBookMarkClickedListener listener;

    public BookMarkRecyclerAdapter(Context context, NotesViewModel viewModel) {
        super(DIFF_CALLBACK);
        this.mContext = context;
        this.mViewModel = viewModel;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_notes, parent, false);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        BookMark currentBookMark = getItem(position);
        holder.mNoteContent.setText(currentBookMark.getNoteContent());
        holder.mNoteTitle.setText(currentBookMark.getNoteTitle());
        holder.mLikeButton.setLiked(true);
        holder.mLikeButton.setOnLikeListener(new OnLikeListener() {
            @Override
            public void liked(LikeButton likeButton) {

            }

            @Override
            public void unLiked(LikeButton likeButton) {
                int noteId = currentBookMark.getNoteId();
                String courseCode = currentBookMark.getCourseCode();
                String noteTitle = currentBookMark.getNoteTitle();
                String noteContent = currentBookMark.getNoteContent();
                BookMark bookMark = new BookMark(noteId, courseCode, noteTitle, noteContent);
                bookMark.setId(currentBookMark.getId());
                mViewModel.deleteBookMark(bookMark);

                SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(mContext.getApplicationContext());
                SharedPreferences.Editor editor = preferences.edit();

                Set<String> defValues = new HashSet<>();
                defValues.add("-1");
                Set<String> noteIds = new HashSet<>(preferences.getStringSet(PreferenceKeys.NOTE_IDS, defValues));
                noteIds.remove(String.valueOf(noteId));

                editor.putStringSet(PreferenceKeys.NOTE_IDS, noteIds);
                editor.apply();
            }
        });
    }

    public BookMark getBookMarkAt(int position) {
        return getItem(position);
    }

    public void setOnBookMarkClickedListener(OnBookMarkClickedListener listener) {
        this.listener = listener;
    }

    public interface OnBookMarkClickedListener {
        void onBookMarkClick(BookMark bookMark);
    }

    public class ViewHolder extends RecyclerView.ViewHolder {

        public final TextView mNoteContent;
        public final TextView mNoteTitle;
        public final LikeButton mLikeButton;

        public ViewHolder(View itemView) {
            super(itemView);
            mNoteContent = itemView.findViewById(R.id.tv_note_content);
            mNoteTitle = itemView.findViewById(R.id.tv_note_title);
            mLikeButton = itemView.findViewById(R.id.likeButton);

            itemView.setOnClickListener(v -> {
                int position = getAdapterPosition();
                if (listener != null && position != RecyclerView.NO_POSITION) {
                    listener.onBookMarkClick(getItem(position));
                }
            });
        }
    }
}

File: app/src/main/java/com/certified/notes/adapters/BookMarkRecyclerAdapter.kt
package com.certified.notes.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.certified.notes.R
import com.certified.notes.model.BookMark

class BookMarkRecyclerAdapter(
//    private val context: Context,
//    private val viewModel: NotesViewModel
) :
    ListAdapter<BookMark, BookMarkRecyclerAdapter.ViewHolder>(
        DIFF_CALLBACK
    ) {

    private lateinit var listener: OnBookMarkClickedListener

//    init {
//        listener = null
//    }

    fun setOnBookMarkClickedListener(listener: OnBookMarkClickedListener) {
        this.listener = listener
    }

    fun getBookMarkAt(position: Int): BookMark {
        return getItem(position)
    }

    interface OnBookMarkClickedListener {
        fun onBookMarkClick(bookmark: BookMark)
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val noteContent: TextView = itemView.findViewById(R.id.tv_note_content)
        val noteTitle: TextView = itemView.findViewById(R.id.tv_note_title)
        val ivBookMark: ImageView = itemView.findViewById(R.id.iv_bookmark)

        init {
            itemView.setOnClickListener {
                val position = absoluteAdapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    listener.onBookMarkClick(getItem(position))
                }
            }
        }
    }

    companion object {
        private val DIFF_CALLBACK: DiffUtil.ItemCallback<BookMark> =
            object : DiffUtil.ItemCallback<BookMark>() {
                override fun areItemsTheSame(oldItem: BookMark, newItem: BookMark) =
                    oldItem.noteId == newItem.noteId

                override fun areContentsTheSame(oldItem: BookMark, newItem: BookMark) =
                    oldItem == newItem
            }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemView =
            LayoutInflater.from(parent.context).inflate(R.layout.list_item_notes, parent, false)
        return ViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val currentBookMark = getItem(position)
        holder.noteContent.text = currentBookMark.noteContent
        holder.noteTitle.text = currentBookMark.noteTitle
        holder.ivBookMark.visibility = View.VISIBLE
//        holder.itemView.setOnClickListener{ onBookMarkClicked?.invoke(currentBookMark)}
//        holder.likeButton.isLiked = true
//        holder.likeButton.setOnLikeListener(object : OnLikeListener {
//            override fun liked(likeButton: LikeButton?) {
//
//            }
//
//            override fun unLiked(likeButton: LikeButton?) {
//                val noteId = currentBookMark.noteId
//                val courseCode = currentBookMark.courseCode
//                val noteTitle = currentBookMark.noteTitle
//                val noteContent = currentBookMark.noteContent
//                val bookMark = BookMark(noteId, courseCode, noteTitle, noteContent)
//                bookMark.id = currentBookMark.id
//                viewModel.deleteBookMark(bookMark)
//
//                val preferences = PreferenceManager.getDefaultSharedPreferences(context)
//                val editor = preferences.edit()
//
//                val defValues = HashSet<String>()
//                defValues.add("-1")
//                val noteIds =
//                    HashSet<String>(preferences.getStringSet(PreferenceKeys.NOTE_IDS, defValues))
//                noteIds.remove(noteId.toString())
//
//                editor.putStringSet(PreferenceKeys.NOTE_IDS, noteIds)
//                editor.apply()
//            }
//        })
    }
}